src/test/fio/fio_ceph_messenger.cc

```diff
@@ -0,0 +1,730 @@
+	io->req_msg->ops.clear();
+
+	/* Here we do not care about direction, always send as write */
+	io->req_msg->write(0, io_u->buflen, buflist);
+	/* Keep message alive */
+	io->req_msg->get();
+	io->req_msg->get_connection()->send_message(io->req_msg);
+
+	return FIO_Q_QUEUED;
+}
+
+static int fio_ceph_msgr_getevents(struct thread_data *td, unsigned int min,
+				   unsigned int max, const struct timespec *ts)
+{
+	struct ceph_msgr_data *data;
+
+	data = (typeof(data))td->io_ops_data;
+
+	/*
+	 * Check io_u.c : if min == 0 -> ts is valid and equal to zero,
+	 * if min != 0 -> ts is NULL.
+	 */
+	assert(!min ^ !ts);
+
+	if (data->io_completed_nr >= min)
+		/* We got something */
+		return min(data->io_completed_nr, max);
+
+	data->lock.Lock();
+	/* Here, under the lock, we are only if min != 0 and ts == NULL */
+	assert(min && !ts);
+
+	while (data->io_completed_nr < min && !td->terminate) {
+		/* Sleep 1 second in order to check early terminate */
+		data->cond.WaitInterval(data->lock, utime_t(1, 0));
+	}
+
+	data->lock.Unlock();
+
+	return min(data->io_completed_nr, max);
+}
+
+static struct io_u *fio_ceph_msgr_event(struct thread_data *td, int event)
+{
+	struct ceph_msgr_data *data;
+	struct ceph_msgr_io *io;
+
+	data = (typeof(data))td->io_ops_data;
+
+	/* TODO: We can avoid locks completely. */
+	data->lock.Lock();
+	assert(data->io_completed_nr);
+	assert(!flist_empty(&data->io_completed_list));
+	io = flist_first_entry(&data->io_completed_list, typeof(*io), list);
+	flist_del(&io->list);
+	data->io_completed_nr--;
+	data->lock.Unlock();
+
+	return io->io_u;
+}
+
+static enum fio_q_status ceph_msgr_receiver_queue(struct thread_data *td,
+						  struct io_u *io_u)
+{
+	struct ceph_msgr_data *data;
+	struct ceph_msgr_io *io;
+
+	io = (typeof(io))io_u->engine_data;
+	data = io->data;
+	data->lock.Lock();
+	if (data->io_pending_nr) {
+		struct ceph_msgr_reply_io *rep_io;
+		MOSDOpReply *rep;
+
+		data->io_pending_nr--;
+		rep_io = flist_first_entry(&data->io_pending_list,
+					   typeof(*rep_io), list);
+		flist_del(&rep_io->list);
+		rep = rep_io->rep;
+		data->lock.Unlock();
+		free(rep_io);
+
+		rep->set_completion_hook(new ReplyCompletion(rep, io));
+		rep->get_connection()->send_message(rep);
+	} else {
+		data->io_inflight_nr++;
+		flist_add_tail(&io->list, &data->io_inflight_list);
+		data->lock.Unlock();
+	}
+
+	return FIO_Q_QUEUED;
+}
+
+static enum fio_q_status fio_ceph_msgr_queue(struct thread_data *td,
+					     struct io_u *io_u)
+{
+	struct ceph_msgr_options *o = (typeof(o))td->eo;
+
+	if (o->is_receiver)
+		return ceph_msgr_receiver_queue(td, io_u);
+	else
+		return ceph_msgr_sender_queue(td, io_u);
+}
+
+static int fio_ceph_msgr_open_file(struct thread_data *td, struct fio_file *f)
+{
+	return 0;
+}
+
+static int fio_ceph_msgr_close_file(struct thread_data *, struct fio_file *)
+{
+	return 0;
+}
+
+/*
+ * Unfortunately old g++ 5.4.0 does not support non-trivial designated
```

Comment:
It can. I do no think using trick with vectors initialization of posval can work out.  And what is wrong with plain old members set?  Of course I can change to any just to make everybody happy, but still :)

Reply:
Again just want a uniform approach in different plugins. Which makes later support easier.